extension-workspace-reference.md

```diff
@@ -0,0 +1,82 @@
+# workspace/reference extension to LSP
+
+The `workspace/reference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references to dependencies that code inside of the workspace makes.
+
+Use case: clients of a language server can invoke `workspace/reference` in order to find external references to dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
```

Comment:
When I studied the protocol for the first time I also noticed that in `ServerCapabilities` they sometimes prepend the namespace, and sometimes they don't. There is a pattern to it though: When there is no other request with the same name in any namespace, they omit the namespace, like `hoverProvider`. When the request exists in multiple namespaces, they prepend it to the capability name, like `documentSymbolProvider` / `workspaceSymbolProvider`. So from regarding that, it should be `workspaceReferenceProvider` here.

As for the prefix, the goal is to not nameclash with the standard version when it makes it into the standard, because the implementation might be slightly different. The `x` stands for extension, like in custom HTTP headers and custom MIME types. It is a prefix, so it should be the first character of the actual property name. For the capability, that would be `xworkspaceReferencesProvider`. Since in PHP I implemented the sub-methods of a namespace as methods of a class, in the files extension I decided to prefix the actual method, so `workspace/xfiles` / `workspace/xreferences`, which looks like this in my implementation:
```php
class LanguageServer {
  /** @var Workspace */
  public $workspace;
  /** @var TextDocument */
  public $textDocument;
}
class Workspace {
  public function symbol() { ... }
  public function xreferences() { ... }
}
```
I would like to keep that style when we extend existing namespaces so I only have to add a new method and not a new `XWorkspace` class. When we add a new namespace, like `cache/get`, we can prefix the whole thing `xcache/get`. Does that make sense?